cdap-docs/developers-manual/source/building-blocks/workflows.rst

```diff
@@ -86,12 +94,114 @@ implement the ``run()`` method::
+logging, modifying the conditional execution, or terminating the execution of workflow.
+
+Execution in the workflow can be made conditional, based on the information contained in
+the token. Execution can be terminated if a node in the workflow produces unexpected
+results.
+
+Once a run is completed, you can query the tokens from past workflow runs for analyses that
+determine which node was executed more frequently and when. You can retrieve the token values
+that were added by a specific node in the workflow to debug the flow of execution.
+
+Scope
+-----
+Two scopes |---| *System* and *User* |---| are provided for workflow keys. CDAP adds keys
+(such as MapReduce counters) under the *System* scope. User programs add their keys under
+the *User* scope.
+
+Putting and Getting Token Values
+--------------------------------
+When a value is put into a token, it is stored under a specific key. Both keys and their
+corresponding values must be non-null. The token stores additional information about the 
+context in which the key is being set, such as the unique name of the workflow node. 
+
+To put a value to a token, first obtain access to the token from the workflow context, and
+then set a value for a specific key, as shown in this example from a MapReduce mapper::
+
+  public static class MyVerifier extends Mapper<LongWritable, Text, Text, NullWritable> {
+    public void map(LongWritable key, Text value, Context context)
+      throws IOException, InterruptedException {
+      String profile;
+      WorkflowToken workflowToken = context.getWorkflowToken();
+      if (workflowToken != null) {
+        if (value != null and value.toString().equals("BuildProductProfile")) {
+          profile = "1";
+        } else {
+          profile = "0";
+        }
+        workflowToken.put("BuildProductProfile", profile);
+      }
+    }
+  }
+
+The `WorkflowToken Java API 
+<../../reference-manual/javadocs/co/cask/cdap/api/workflow/WorkflowToken.html)>`__
+includes methods for getting values for different keys, scopes, and nodes. The same
+key can be added to the workflow by different nodes, and there are methods to return a map of those
+key-value pairs. Convenience methods allow the putting and getting of non-string values
+through the use of the class Value.
+
+Persisting the WorkflowToken
+----------------------------
+
+The RunRecord for the workflow contains the WorkflowToken as a property. This token
+is persisted after each action completes in a Workflow. 
+
+Examples
+--------
+
+In this code sample, we show how to update the WorkflowToken in a MapReduce program::
+
+  @Override
+  public void beforeSubmit(MapReduceContext context) throws Exception {
+    ...
+    WorkflowToken workflowToken = context.getWorkflowToken();
+    if (workflowToken != null) {
+      // Put the action type in the WorkflowToken
+      workflowToken.put("action_type", "MAPREDUCE");
```

Comment:
Sorry, my bad, I should have updated document to reflect this. 
Its better to put the keys by `.` separated, instead of `_`. so the key will be `action.type`.